### PR TITLE
[client] Highlight linked comments to identify them

### LINKED ISSUE #838: [client] When scrolling to a linked comment, it should have some highlight
When using a link such as `https://example.com/post-1/#isso-4` for example, the comment that is being linked to should flash a couple times to make it identifiable from other nearby comments.

### PR DESCRIPTION
**Pull request by BBaoVanC**
Fixes #838 

TODO:

- [x] Figure out the best way to do styling
- [x] Docs
- [x] Tests (check that `.isso-target` gets applied properly)

### CODE REVIEW

**Comment by BBaoVanC** [comment]
I meant to make this a WIP pull request, but forgot to before submitting.

**Comment by BBaoVanC** [comment]
@ix5 Can you review this please?

I have CSS added to set a background color included. I don't think that separating it out like what we did in #844 makes sense because this feature should be included by default. It's also different since it doesn't require any configuration by the site owner to function.

Also, I did figure out how to get it not to conflict with #844. By not setting the 100% keyframe, it automatically just fades out to whatever background color is the fallback for the element.

Let me know if there's a better place for the `hashchange` event listener, and if `window.addEventListener` is the right function to use for it; I wasn't really sure what to do with it.

**Comment by ix5** [comment]
Hey @BBaoVanC, I'm off to a well-deserved break, starting tomorrow. So I won't be able to review your changes for about a week.

Let's do a deal: You figure out a solution and submit draft PRs to at least some of these open issues which I think you're more than capable of solving:
- https://github.com/posativ/isso/issues/111
- https://github.com/posativ/isso/issues/448
- https://github.com/posativ/isso/issues/465
- https://github.com/posativ/isso/issues/859
- Race conditions / re-arranging parts in https://github.com/posativ/isso/pull/821 and https://github.com/posativ/isso/issues/842

... and I'll be happy to take some time to review this highlighting PR once I get back ;-)

Would also be nice if at some of these low hanging fruits could be fixed:
- https://github.com/posativ/isso/issues/552
- https://github.com/posativ/isso/issues/771

**Comment by BBaoVanC** [comment]
> Hey @BBaoVanC, I'm off to a well-deserved break, starting tomorrow. So I won't be able to review your changes for about a week.
> 
> Let's do a deal: You figure out a solution and submit draft PRs to at least some of these open issues which I think you're more than capable of solving:
> 
>     * [Make name/URL/e-mail editable #111](https://github.com/posativ/isso/issues/111)
> 
>     * [Counter problem #448](https://github.com/posativ/isso/issues/448)
> 
>     * [Bugs with indented code #465](https://github.com/posativ/isso/issues/465)
> 
>     * [[client] Clean up and standardize client CSS #859](https://github.com/posativ/isso/issues/859)
> 
>     * Race conditions / re-arranging parts in [[client] js: Fetch public conf once from /config; embed.js: Preserve default config values #821](https://github.com/posativ/isso/pull/821) and [[client] `fetchComments` is adding another PostBox even if there is already one displayed #842](https://github.com/posativ/isso/issues/842)
> 
> 
> ... and I'll be happy to take some time to review this highlighting PR once I get back ;-)
> 
> Would also be nice if at some of these low hanging fruits could be fixed:
> 
>     * [Log out of admin interface #552](https://github.com/posativ/isso/issues/552)
> 
>     * [Isso always calls count even if not needed #771](https://github.com/posativ/isso/issues/771)

I'll take a look at those. Hope you have a nice break!

File: isso/css/isso.css [hunk 1]
```diff
@@ -159,6 +159,13 @@
     display: none;
 }
 
+.isso-target {
+    animation: target-fade 5s ease-out;
+}
+@keyframes target-fade {
```
Comment (ix5) [inline isso/css/isso.css:165]: That animation name should be isso-namespaced

File: isso/css/isso.css [hunk 2]
```diff
@@ -159,6 +159,13 @@
...
+    0% { background-color: #eee5a1; }
```
Comment (ix5) [inline isso/css/isso.css:166]: This color (light yellow) is quite jarring on dark background. Try e.g. the demo with `body { background: black; color: white; }`

Also, maybe add a little code comment such as `Flash comment light yellow when focussed`, either here or in embed.js

File: isso/js/embed.js [hunk 3]
```diff
@@ -105,6 +112,9 @@ function fetchComments() {
             if (window.location.hash.length > 0 &&
                 window.location.hash.match("^#isso-[0-9]+$")) {
                 $(window.location.hash).scrollIntoView();
```
Comment (ix5) [inline isso/js/embed.js:114]: Wrap this in a try-catch block (ES5 has it, I think?) or check if `$(window.location.hash` actually returns an element. `fetchComments` (and `init`) is a function that definitely should not throw.

File: isso/js/embed.js [hunk 4]
```diff
@@ -105,6 +112,9 @@ function fetchComments() {
...
+
+                // We can't just set the id because that would overwrite the existing `isso-X` id
```
Comment (ix5) [inline isso/js/embed.js:116]: Could you elaborate on what you meant by that code comment?

**Review by ix5 — COMMENTED** [review]
> I have CSS added to set a background color included. I don't think that separating it out like what we did in #844 makes sense because this feature should be included by default. It's also different since it doesn't require any configuration by the site owner to function.

That's fine. As commented inline, the yellow is a bit jarring at night, but yellow is universally accepted as a marker/highlight/attention color. Maybe once there's a dark mode someone can style it differently. Keep as-is for now.

> Let me know if there's a better place for the hashchange event listener, and if window.addEventListener is the right function to use for it; I wasn't really sure what to do with it.

I'm not qualified to comment on event listeners, my JS knowledge is pretty bad. I found [this SO answer](https://stackoverflow.com/questions/6390341/how-to-detect-if-url-has-changed-after-hash-in-javascript) which seems to suggest a similar thing. Availability of [Caniuse: hashchange](https://caniuse.com/hashchange) seems good.

As for stuffing this piece of code into `init()`: IMO that file is already quite overloaded. Maybe shoving everything related to timers and events into a separate file (that can be stubbed/mocked) would simplify testing a good bit.

Several places rely on `window.location.*`. Unifying that would be a nice first step.
E.g. this is just a stupid duplication:
https://github.com/posativ/isso/blob/fdcadaddcd87fe97fbd7ca328a2ef7ea56eccb87/isso/js/app/api.js#L7
https://github.com/posativ/isso/blob/fdcadaddcd87fe97fbd7ca328a2ef7ea56eccb87/isso/js/embed.js#L74

File: isso/js/embed.js (hunk 3 above)
Comment (BBaoVanC) [reply isso/js/embed.js:114]: Yep, can't believe I didn't catch this error when I was testing. (It happens if you click a comment for the first time, before another one was linked, meaning there is no .isso-target)

**Comment by ix5** [comment]
Would you mind adding a small styling guide to the docs? (Perhaps under "Tips & Tricks", 1-2 lines should be enough).

Code-wise, this looks good right now.

**Comment by BBaoVanC** [comment]
> Would you mind adding a small styling guide to the docs? (Perhaps under "Tips & Tricks", 1-2 lines should be enough).
> 
> Code-wise, this looks good right now.

I added some docs, now all that's left is a small test if possible which just makes sure that `.isso-target` gets applied when going to like `#isso-1` for example.

**Comment by ix5** [comment]
Added some tests. Thank you and looking forward to reviewing other great contributions!

**Pull request merged**